Storage Add Ons:  Implement dropdown checkout functionality

jeyip opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/wp-calypso/pull/79041#discussion_r1279224891
Issue https://github.com/Automattic/martech/issues/2025
Tracking https://github.com/Automattic/martech/issues/2023

## Proposed Changes

* Update checkout cart for `/start` signup onboarding flow with storage add ons

## GIFs
![2023-08-31 16 10 43](https://github.com/Automattic/wp-calypso/assets/5414230/157f43f0-6f5c-458c-b822-93fb922cf61d)

## To be left as a Follow Up for the Future
* Make the storage add-on in checkout cart a dropdown ( to allow user to switch between storage add-ons ) @saygunnyc has stated that this is something we can build on iteratively

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Smoke test flows:  in PCYsg-RDT-p2 and ensure that a free plan ( if available ) and a paid plan can be purchased
  * /setup/newsletter ( Stepper flows should be unaffected by our changes )
    * navigate to http://calypso.localhost:3000/setup/newsletter as a logged in user
    * Complete onboarding steps until you’ve reached the plans page 
  * /start/domains
    * Navigate to http://calypso.localhost:3000/start/domains
    * Select a custom domain like .org, .blog, etc.
    * Navigate to the plans page
  * /start/hosting
    * Navigate to http://calypso.localhost:3000/start/hosting
    * Navigate to the plans page
  * /start/import
    * Navigate to http://calypso.localhost:3000/start/import
    * Select any domain, free or custom
    * Navigate to the plans page
  * /start/onboarding-pm
    * Navigate to http://calypso.localhost:3000/start/onboarding-pm
    * Select any domain, free or custom
    * Navigate to the plans page
  * /start/onboarding-with-email
    * Navigate to http://calypso.localhost:3000/start/onboarding-with-email
    * Select any domain, free or custom
    * Select the provided email option
    * Navigate to the plans page
  * /start/with-themes
    * Go to [https://wordpress.com/themes while](http://calypso.localhost:3000/themes%C2%A0while) logged out and pick a Premium theme.
    * Alternatively, copy the URL of the “Pick this design” button and paste it in a logged-in browser session (it should be something like wordpress.com/start/with-theme?theme=yuga&premium=true&theme_type=premium)
    * Proceed with the onboarding until you get to the “Choose a plan” step
  * /start/personal
    * Navigate to http://calypso.localhost:3000/start/personal
    * Select any domain, free or custom
  * /start/personal-monthly
    * Navigate to http://calypso.localhost:3000/start/personal
    * Select any domain, free or custom
  * /start/personal-2y
    * Navigate to http://calypso.localhost:3000/start/personal
    * Select any domain, free or custom
  * /start/business
    * Navigate to http://calypso.localhost:3000/start/business
    * Select any domain, free or custom
  * /setup/wooexpress ( since wooexpress has no onboarding plans page, it should be unaffected by our changes )
    * Navigate to http://calypso.localhost:3000/setup/wooexpress ( If testing in a dev environment, make sure to navigate to the wordpress.com URL above to create the initial site. Then switch over to a dev URL. For some reason, using a dev URL to initially set up wooexpress doesn’t work as expected )
    * Wait until redirected to the wp-admin page for the newly created atomic site
    * From the wp-admin page click, on Upgrades > Plans
    * Alternatively, navigate to http://calypso.localhost:3000/plans/{SITE_SLUG}

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

jeyip committed:
Add dropdown functionality to storage add-on upsell

jeyip commented on client/landing/stepper/declarative-flow/internals/steps-repository/plans/plans-wrapper.tsx:
```diff
@@ -75,7 +75,8 @@ const PlansWrapper: React.FC< Props > = ( props ) => {
 	}, [] );
```
**Note:**

Amendments to this file allow us to add storage add-ons to the shopping cart in stepper onboarding flows

jeyip commented on client/lib/signup/step-actions/index.js:
```diff
@@ -1270,7 +1269,7 @@ export function isNewOrExistingSiteFulfilled( stepName, defaultDependencies, nex
 }
 
-export const buildUpgradeFunction = ( planProps, cartItem ) => {
+export const buildUpgradeFunction = ( planProps, cartItem, addOnItem ) => {
```
**Note:**

Amendments to this file allow us to add storage add-ons to the shopping cart in **signup** onboarding flows

jeyip commented on client/my-sites/plans-features-main/index.tsx:
```diff
@@ -295,15 +301,27 @@ const PlansFeaturesMain = ( {
 			? getPlanPath( cartItemForPlan.product_slug )
 			: '';
-		const checkoutUrlWithArgs = `/checkout/${ siteSlug }/${ planPath }`;
+		const checkoutUrlWithArgs = `/checkout/${ siteSlug }/${ planPath },${ storageAddOn?.productSlug }:-q-${ storageAddOn?.quantity }`;
```
**Note:**

This change adds storage add-ons to the checkout cart for the `/plans` page

jeyip commented:
**Note:**

I'm seeing some strange caching issues for add-ons data when visiting the `/plans` page. I'm currently investigating, but test may not work in `/plans` as expected 🤔

jeyip commented:
**Note:**

There's a delay between when a storage upgrade add-on is purchased and when the purchase is reflected in the `/plans` page ( it will look as if nothing has changed until a minute or two later ). I'll work more on this tomorrow

jeyip commented on client/my-sites/plans/main.jsx:
```diff
@@ -411,6 +412,7 @@ class Plans extends Component {
 				{ selectedSite.ID && <QuerySitePurchases siteId={ selectedSite.ID } /> }
 				<DocumentHead title={ translate( 'Plans', { textOnly: true } ) } />
 				<PageViewTracker path="/plans/:site" title="Plans" />
+				<QueryProductsList />
```
The products list is necessary to load add-ons data properly, which we now require in the `/plans` page

southp commented on client/my-sites/add-ons/hooks/use-add-ons.ts:
```diff
@@ -140,7 +140,7 @@ const useAddOns = ( siteId?: number ): ( AddOnMeta | null )[] => {
 		const spaceUpgradesPurchased: number[] = [];
 
-		if ( filteredTransactions?.length ) {
+		if ( filteredTransactions?.length && siteId ) {
```
I'm not sure I understand why the siteId all of a sudden is needed here. Could you help me understand?

jeyip replied:
Thanks for the patience! 

This was meant to fix a bug when we call `use-add-ons` in the context of onboarding. For some reason, add-on data incorrectly decides that both 50GB add-ons and 100GB add-ons are _already_ purchased, even though no site has been created yet.

<img width="1487" alt="Screenshot 2023-09-15 at 1 29 29 AM" src="https://github.com/Automattic/wp-calypso/assets/5414230/48651b84-9520-4504-a9fa-9de16ad19adc">

To elaborate on why, `filteredTransactions` returns transactions for a given site when `siteId` exists. When, however, siteId is undefined, `filteredTransactions` fetches _all_ transactions for a user, regardless of what site the purchase was made on. 

Since I had made _some_ space upgrade add-on purchase in the past on _any_ site, the logic then proceeded to mark add-on data as `purchased`

https://github.com/Automattic/wp-calypso/blob/426d78e95c58f31f2d1f9c19ade66e4dd33a6504/client/my-sites/add-ons/hooks/use-add-ons.ts#L231-L244

By checking for a truthy `siteId`, I intended to only look for space upgrade add-on transactions if the site has already been created ( and we're in the plans page )

With that being said, I wrote this when I was _really_ tired 😴 , so we might consider better inline documentation or exploring alternative ways to address the bug. Ex. It looked like it might be possible to update the params for `filterTransactions` such that, when called with an undefined siteId, it returns no transactions, rather than all transactions, but I haven't looked to closely yet

southp replied:
Thanks for the explanation. It looks like the transaction data is used for filtering out the storage add-ons that have lower volumn than the owned one. Also, how the whole transaction data is fetched and filtered is actually quite non-ideal –– it fetches the whole transaction history and than filter from the client side, while all these should be done from the server side so our users don't have to waste their bandwidth and local computational resources to perform all these. Since correcting the whole thing would be out of scope, let's start by not fetching those transaction data in the onboarding flow since that doesn't make sense. It'd involve some level of decoupling, but we will have a cleaner code without tricky conditions needing explicit documentation :)

jeyip replied:
I took a very simplistic approach to this in https://github.com/Automattic/wp-calypso/pull/81340/commits/ce39121d8eb4f2da5bf81c2d41e0ba369b0f9621 and  https://github.com/Automattic/wp-calypso/pull/81340/commits/21e2cfbaa34ac0ba34e05be06c4045d3e1d1aa4e. I added a guard with `isInSignup` to prevent fetching. I'm not sure if that's the best flag, but it seemed reasonable to me

southp reviewed (commented):
@jeyip I didn't manage to look into the code deeply today, but the `/start` flow works as expected :) 

Since we are now restricting it into `/start` only, could you help remove all the unncessary part outside of that? That will greatly reduce the required amout of code review and testing :)

jeyip commented:
> I didn't manage to look into the code deeply today, but the /start flow works as expected :)

> Since we are now restricting it into /start only, could you help remove all the unncessary part outside of that? That will greatly reduce the required amout of code review and testing :)

Working on it now :D

jeyip commented on client/signup/config/steps-pure.js:
```diff
@@ -117,8 +117,8 @@ export function generateSteps( {
 			dependencies: [ 'siteSlug' ],
-			providesDependencies: [ 'cartItem', 'themeSlugWithRepo' ],
-			optionalDependencies: [ 'themeSlugWithRepo' ],
+			providesDependencies: [ 'cartItem', 'themeSlugWithRepo', 'addOnItem' ],
```
**Note:**

We have to specify add-on items for this many steps because we want to add a new dependency for `plans`, which has a number of linked steps. We can see a sample of them below: https://github.com/Automattic/wp-calypso/blob/c33f1f6be415ba7adf63c2995eadd0cda4081e68/client/signup/config/step-components.js#L28-L42

More about bugs with linked steps and unspecified dependencies can be found here https://github.com/Automattic/wp-calypso/blob/c33f1f6be415ba7adf63c2995eadd0cda4081e68/client/signup/config/test/index.js#L23-L36

jeyip commented:
@vinimotaa 👋 

Hey Vini! After purchasing the space storage add-on, it takes a minute or two for all of the information to be correctly updated in the database ( Ex. I purchase the 50GB add-on, so the total storage for the site goes from 50GB to 100GB ). 

When a user visits the plan page while the data is still being updated, do you have thoughts on what we should display for the storage label to indicate the pending state?

southp commented:
@jeyip Sorry for being ignorant about the design decisions made in the past, but could you help me point out the reference of showing the standard storage + the add-on storage in the plans grid? I'm asking since I feel this is a bit confusing:

1. If we want to show the accumulated space, then I'm wondering if we should also do that for Free / Personal / Premium. Nothing prevent them from purchasing storage add-ons anyway :) 
2. Overall I feel it's risky to show the accumulated space without further explanation. As someone who is familar with that label, I might think the eCommerce plan has 100G space so I'd have 100G + 50G space once upgraded. Have we considered to separate them as two labels or something if we really want to emphasize the extra space from the owning storage add-on in the plans grid?

Also cc @gmovr and @dzver here in case you both happen to know this on top of your mind :)

dzver commented:
@southp I don't know on top of my head. My goal is to unblock this task so that it gets shipped, and I'm leaning toward simplification rather than making things more complex. Whatever the design is, it has to show we support space upgrades, and that's pretty much the only solid requirement.

@jeyip could you point to the latest design on this? I don't see a link in the PR or the corresponding issue.

jeyip commented:
> Sorry for being ignorant about the design decisions made in the past, but could you help me point out the reference of showing the standard storage + the add-on storage in the plans grid?

Sure I'd be happy to:
- p9Jlb4-7Ui-p2
- https://github.com/Automattic/martech/issues/1907

southp commented:
Thanks for pointing out the references for me, @jeyip :) 

From my view, this design decision works with the dropdowns, since the label and the dropdowns are obviously communicating different things. However, when all that we have is the label, it doesn't work since they look the same but having different meanings. If we still agree to the original decision of rolling the add-on UI out to `/start` only, I think it's the best to not touch the label and just leave `/plans` the way it is –– we already have the storage add-ons and it's been this way, so it shouldn't be a deal breaker. We will then work on `/plans` as the next milestone. 

Does that make sense to you?

jeyip commented:
> If we still agree to the original decision of rolling the add-on UI out to /start only, I think it's the best to not touch the label and just leave /plans the way it is –– we already have the storage add-ons and it's been this way, so it shouldn't be a deal breaker. We will then work on /plans as the next milestone.

> Does that make sense to you?

This does! If we end up changing our minds, we can draft a relatively quick PR to update the storage add-ons label in `/plans` as well

jeyip committed:
Update plans-ecommerce-fulfilled step

jeyip committed:
Always show default storage option slug for plan

As it exists today, storage labels show the default amount of storage in the pricing grid, even if a storage upgrade add-on is purchased. We ensure that we continue to do so if the storage add-on dropdown is not displayed

jeyip commented on client/my-sites/plan-features-2023-grid/hooks/npm-ready/data-store/use-grid-plans.tsx:
```diff
@@ -71,12 +71,10 @@ export type UsePricingMetaForGridPlans = ( {
 	planSlugs,
 	withoutProRatedCredits,
 	storageAddOns,
-	currencyCode,
```
**Note:**

currencyCode was an unused argument

jeyip commented:
@southp Rebased, rolled back any storage label changes to `/plans`, and responded to all comments 😄 This PR is ready for review

chriskmnds commented on client/my-sites/plans-features-main/index.tsx:
```diff
@@ -302,8 +308,20 @@ const PlansFeaturesMain = ( {
+
 		if ( onUpgradeClick ) {
-			onUpgradeClick( cartItemForPlan );
+			onUpgradeClick( cartItemForPlan, storageAddOnItemForPlan );
```
Why is `storageAddOnItemForPlan` treated differently from `cartItemForPlan`? Is the add-on _not_ a cart item, and it needs to be sent separately?

jeyip replied:
Good observation -- we'd have to update a relatively foundational method in order to handle multiple cart items at once ( [buildUpgradeFunction](https://github.com/Automattic/wp-calypso/blob/trunk/client/lib/signup/step-actions/index.js#L1273-L1284) ). This will improve code quality. From what I can tell, the surface are of code changes would also be small. It will, however, probably require extensively testing each signup flow to make sure everything continues to function as expected

We'd want to update:

`onUpgradeClick` -> `onSelectPlan` -> `buildUpgradeFunction`

@southp just wanted to keep you in the loop on this given our deadlines. If we move forward, it will probably delay delivery by a day to three days ( to account for extra testing and code review ). The alternative is to create a follow-up that we could tackle at a different time, but I'm more than happy to work on this if we're all in agreement 🙂

southp replied:
It's a valid concern worth addressing. From the other exsiting step action [addAddOnsToCart()](https://github.com/Automattic/wp-calypso/blob/trunk/client/lib/signup/step-actions/index.js#L638) we clearly need to make the data pipeline filling the shopping cart capable of handling multiple items. Let's give it a shot and I'll be happy to assist in testing and buying us more time if needed. 1~3 days doesn't sound too bad for me given what we'd gain from here :)

chriskmnds replied:
Cool. Happy to assist with anything needed too /review, testing, etc. If we can unify the concept without unhealthy dose of complexity or side effects (like querying plans-ui store outside of plans code), then it sounds like good direction :-)

> like querying plans-ui store outside of plans code

Saying this without investigating more. It _sounds_ like something we can avoid 🤔

jeyip replied:
Wonderful! Thanks for reaching a consensus with me everyone -- I believe I'll have something to address the cart items shortly 😃

jeyip replied:
Hey folks! You might see checkout code being pushed in https://github.com/Automattic/wp-calypso/pull/81340/commits/9d67605a29d348e9f37ae26fb4a999dd0c09b2c6 -- the functionality appears to work, but to be clear, I'm still tidying things up 🙂

If there are glaring issues with the general direction though, feel free to chime in

jeyip committed:
Support multiple checkout cart items in buildUpgradeFunction

jeyip commented on client/lib/signup/step-actions/index.js:
```diff
@@ -606,9 +607,9 @@ async function addExternalManagedThemeToCart( state, dispatch, themeId, siteSlug
-	if ( isEmpty( cartItem ) && isEmpty( emailItem ) ) {
+	const { emailItem, siteSlug, plugin, billing_period: billingPeriod } = dependencies;
+	const { cartItems, lastKnownFlow } = stepProvidedItems;
+	if ( cartItems.some( ( item ) => item === null ) && isEmpty( emailItem ) ) {
```
**Note:**

I'm not using isEmpty because of broader efforts to deprecate `lodash` https://calypsop2.wordpress.com/2021/02/24/the-path-to-deprecating-lodash/

jeyip committed:
Fix various bugs and units tests

Revert isEmpty change

Remove unnecessary check for cartItem

Fix pro plan upgrade button

Fix plan step unit test

Fix typescript errors

jeyip commented:
Currently:

- Fixing a failing unit test
- Cleaning up code that supports multiple cart items for the `buildUpgradeFunction`
- Working on James's suggestion https://github.com/Automattic/wp-calypso/pull/81340#discussion_r1329501378

jeyip commented on client/lib/signup/step-actions/index.js:
```diff
@@ -606,9 +612,9 @@ async function addExternalManagedThemeToCart( state, dispatch, themeId, siteSlug
-	if ( isEmpty( cartItem ) && isEmpty( emailItem ) ) {
+	const { emailItem, siteSlug, plugin, billing_period: billingPeriod } = dependencies;
+	const { cartItems, lastKnownFlow } = stepProvidedItems;
+	if ( cartItems === null && isEmpty( emailItem ) ) {
```
Note:

I'm not using isEmpty because of broader efforts to deprecate lodash p4TIVU-9Bf-p2

southp replied:
It's great but `cartItems === null` isn't equivalent to `isEmpty()`. From my rough understanding to the code we'd also need to check if cartItems is an empty array, but I'm not sure. If that's the case, just using `isEmpty()` would be easier, since the rest of the file is using it anyway :)

jeyip replied:
> From my rough understanding to the code we'd also need to check if cartItems is an empty array, but I'm not sure.

Since we're only dealing with plans cart items in the `addPlanToCart` step action, from what I've seen so far, `cartItems` will not be an empty array. It's part of the reason I actually switched to an explicit null check instead. I believe you're mentioning this because of the failing tests https://github.com/Automattic/wp-calypso/pull/81340#issuecomment-1727497913?

I need to double check, but my understanding of the test failure was something different. It seemed like free plans were only represented as `null` values, but the test has a non-nullish free plan cart item it passes as an argument. Have both a `null` and a non-null free plan seems extremely confusing 😬

https://github.com/Automattic/wp-calypso/blob/d102dd7759f0aa04a81563ae8b9a6525d8f8efb6/client/signup/steps/plans/test/index.jsx#L125-L130

With all of that being said, I believe `isEmpty` should still work as before so I'll switch it back 👍

jeyip replied:
https://github.com/Automattic/wp-calypso/pull/81340/commits/1edf6b965b53f0e9a294eb19953df2ccd68633d

southp commented on client/lib/signup/flow-controller.ts:
```diff
@@ -442,7 +444,15 @@ export default class SignupFlowController {
+				!! progress.plans?.cartItem ||
+				progress.plans?.cartItems?.find( ( item: MinimalRequestCartProduct ) => isPlan( item ) )
+			);
+		}
```
I'm wondering if we really have to check whether there is a plan in `cartItems`. Shouldn't we always need to go through checkout if there is anything in the cart? 🤔

jeyip replied:
Ah this discussion was part of the todos I wanted to get to yesterday but didn't have time for https://github.com/Automattic/wp-calypso/pull/81340#discussion_r1330767937. We briefly touched on this in our chat yesterday

The answer to your question is yes, and there is _another_  check for generic `cartItems` here https://github.com/Automattic/wp-calypso/pull/81340/commits/9d67605a29d348e9f37ae26fb4a999dd0c09b2c6#diff-a7bd84bdbc16a9cd2414a1ac5c5d990a00617b9f9a6c678a867e1257646b9385R34-R36.

That method is eventually called by filterDestination, which, like `_getNeedsToGoThroughCheckout` _also_ decides whether or not to redirect to checkout 😵 I haven't dug into why two checks exist

https://github.com/Automattic/wp-calypso/blob/48ab924b8b885ded3cfdee44b736d4b6e6f9fe01/client/signup/config/flows.js#L278-L281

I was planning to create an issue for this because it feels out of scope, and, in my opinion also seems like it might be a rabbit hole to investigate and fix now ( there's been quite a lot of those in the signup codebase ). Open to hearing your thoughts though 🙂

southp commented:
Thanks for working the `cartItems` out, @jeyip . It should be very close once all the failing tests are fixed. I didn't manage to fix some of them as I originally planned to, but my impression is that `[]` and `null` is representing the same or overlapping meaning now, giving us quite some headache to manage. The signup framework doesn't seem to provide a way to default a dependency value(or does it? @aneeshd16 @jdc91 do you recall?). Otherwise it might be easier.

chriskmnds commented on client/my-sites/plans-features-main/index.tsx:
```diff
@@ -302,8 +305,24 @@ const PlansFeaturesMain = ( {
 		}
 
+		const selectedStorageOption =
+			selectedStorageOptions && selectedStorageOptions[ cartItemForPlan?.product_slug || '' ];
+		const storageAddOn = storageAddOns?.find( ( addOn ) => {
+			return addOn?.featureSlugs?.includes( selectedStorageOption || '' );
+		} );
+
+		const storageAddOnItemForPlan = storageAddOn && {
+			product_slug: storageAddOn.productSlug,
+			quantity: storageAddOn.quantity,
+			volume: 1,
+		};
+
 		if ( onUpgradeClick ) {
-			onUpgradeClick( cartItemForPlan );
+			const cartItems = [
+				...( cartItemForPlan ? [ cartItemForPlan ] : [] ),
+				...( storageAddOnItemForPlan ? [ storageAddOnItemForPlan ] : [] ),
+			];
```
This logic feels a little disconnected to me. I think either none of it should happen here and instead go into the `handleUpgradeClick` of the grid components, OR unify everything (plan & add-on selection) to go through the store.

I'd have expected these to be unified:

1. user interacts with grid and store gets updated with selections
2. on plan upgrade, the elements are put together sent through the callback here for processing OR both selected plan and add-ons are accessed through the data storehere

VS

1. user interacts with grid and store gets updated with the selections
2. on plan upgrade, plan is sent through as argument and wrapper (`plans-features-main`) has to query the store for the selected add-ons (and put together the respective object)

The side effect is `plans-features-main` accessing the plans-ui data store. I'd rather if we either remove the side-effect or unify everything behind it.

chriskmnds replied:
obviously opinion. might be a different angle to it

jeyip replied:
> This logic feels a little disconnected to me. I think either none of it should happen here and instead go into the handleUpgradeClick of the grid components, OR unify everything (plan & add-on selection) to go through the store.

Hi @chriskmnds! Thanks for bringing this up -- to be honest, I spent so much time wrapping my head around signup plans code yesterday and fixing related bugs that I forgot about your comment here https://github.com/Automattic/wp-calypso/pull/81340#discussion_r1329965409. It's my mistake. I'm considering  your review now 🤔

jeyip replied:
> The side effect is plans-features-main accessing the plans-ui data store. I'd rather if we either remove the side-effect or unify everything behind it.

> I think either none of it should happen here and instead go into the handleUpgradeClick of the grid components, OR unify everything (plan & add-on selection) to go through the store.

> on plan upgrade, the elements are put together sent through the callback here for processing OR both selected plan and add-ons are accessed through the data storehere

Hey @chriskmnds  -- I think I understand what you're trying to explain, but there might be a few points of confusion on my end that I'd like to clarify. Here's what I understand about what you're trying to convey:

- You're suggesting that we _either_ move logic to build `cartItems` from the `handleUpgradeClick` in `plans-features-main` into the `handleUpgradeClick` in `plan-features-2023-grid`, or handle _both_ plan and add-on selection with the `wpcom-ui-store` in `plans-features-main`. 
- The option to use `wpcom-ui-store` for both plan and add-on selection makes sense to me, in that it means that there wouldn't be two forms of state management being consolidated into `handleUpgradeClick` in `plans-features-main`
- The confusion for me is in the option to migrate logic into `handleUpgradeClick` in `plan-features-2023-grid`. Are we attempting to unify state management here as well? Are you also saying that, once we've moved the logic into `plan-features-2023-grid`, we'd want to use local state management to handle add-ons selection alongside plan selection? This is something that seems like it'd be a step backwards if that's the case. 🤔
- Or is the side effect with `plans-features-main` the main concern? Perhaps because we want the `wpcom-ui-store` to be used exclusively with the `plan-features-2023-grid`? And moving the existing side effect from `plans-features-main` to `plan-features-2023-grid` is enough?

jeyip committed:
Don't fetch add-on transaction history in onboarding

jeyip commented:
Currently:

- Working thinking about and working on Christo's suggestion https://github.com/Automattic/wp-calypso/pull/81340#discussion_r1331477865
- Continuing to clean up and refactor `cartItems` checkout code

jeyip committed:
Build cartItems in the plan features grid

southp commented:
Just want to jot down the test cases I'll be going through when the time comes:

### Classic signup
Create a free site/create a paid site/create a paid site + a paid domain should work as expected on the following flows:

- [ ] /start
    - [ ] try out different combinations with different add-ons
- [ ] /start/domain
- [ ] /start/do-it-for-me
- [ ] /start/p2

Plan specific flows should work intact:

- [ ] /start/personal
- [ ] /start/premium
- [ ] /start/business
- [ ] /start/ecommerce

### Stepper 
They shouldn't be affected, but worth a check.

Create a free site/create a paid site/create a paid site + a paid domain should work as expected on the following flows:
- [ ] /setup/newsletter
- [ ] /setup/link-in-bio
- [ ] /setup/hosting-site
- [ ] /setup/design-first

Flows that don't have a plan step should work intact

- [ ] /setup/free
- [ ] /setup/videopress
- [ ] Woo Express

### Logged-in

- [ ] /plans: should be in general functional as it is in production
- [ ] /plans: upgrade to the higher-tier plans should work as expected
- [ ] launch site flow: launch as the free site
- [ ] launch site flow: launch as a paid site
- [ ] launch site flow: launch as a paid + domain site

chriskmnds commented on client/my-sites/plan-features-2023-grid/index.tsx:
```diff
@@ -55,7 +57,7 @@ import type {
 } from './hooks/npm-ready/data-store/use-grid-plans';
 import type { PlanActionOverrides } from './types';
-import type { DomainSuggestion } from '@automattic/data-stores';
+import type { DomainSuggestion, selectedStorageOptionForPlans } from '@automattic/data-stores';
```
Types should always be ~UpperCased~ capitalised. `selectedStorageOptionForPlans` -> `SelectedStorageOptionForPlans`

jeyip replied:
Resolved in https://github.com/Automattic/wp-calypso/pull/81340/commits/f7e6cc6e5951875fc9ccd1dac8ae6ddf5b52bbe0

chriskmnds commented on client/my-sites/plan-features-2023-grid/index.tsx:
```diff
@@ -440,14 +444,32 @@ export class PlanFeatures2023Grid extends Component< PlanFeatures2023GridType >
 
-		// TODO clk: Revisit. Could this suffice: `ownPropsOnUpgradeClick?.( cartItemForPlan )`
+		const selectedStorageOption =
+			selectedStorageOptions && selectedStorageOptions[ cartItemForPlan?.product_slug || '' ];
```
In general, I think it's best if we don't retrieve things from data structures using fallback empty values like this. It confuses further reasoning with the type. In this case, `selectedStorageOption` is always defined despite `''` being used as the effective value.

However, this can be simplified to:

```
const selectedStorageOption = selectedStorageOptions?.[ planSlug ];
```

Otherwise, maybe consider a more deterministic expression?:

```
const selectedStorageOption = cartItemForPlan?.product_slug ? selectedStorageOptions?.[ cartItemForPlan?.product_slug ] : undefined;
```

jeyip replied:
Thank you! Makes total sense. I'll roll this into the other code cleanup I wanted to wrap up this morning 👍

jeyip replied:
Tidied up your suggestion and a subsequent empty value comparison https://github.com/Automattic/wp-calypso/pull/81340/commits/456b5d64d4fff73aa1e48e3a21f46fc0c4881c4f

chriskmnds reviewed (commented):
Thanks for following up. I think you brought it to the right place now. 🎉 
We can later turn the handler into a callback hook and reuse across Comparison and Features Grid (when split/extracted).

Just some comments below while going through the code

chriskmnds commented on client/my-sites/add-ons/hooks/use-add-ons.ts:
```diff
@@ -130,22 +130,25 @@ const useAddOns = ( siteId?: number ): ( AddOnMeta | null )[] => {
+		// creating a brand new site. For the time being, skip fetching transaction history in onboarding
+		// flows.
+		if ( billingTransactions && ! isInSignup ) {
+			const filteredTransactions = filterTransactions( billingTransactions, filter, siteId );
```
If changing the block scope for `filteredTransactions`, then do you need to do the same for `filter` from above?

jeyip replied:
Ah yes -- and I'm also now seeing that I missed one or two more issues. I'll fix them shortly. Thank you!

jeyip replied:
Resolved in https://github.com/Automattic/wp-calypso/pull/81340/commits/c51a366d33a7c4648373700a3c9b63086002452a

chriskmnds commented on client/lib/signup/step-actions/index.js:
```diff
@@ -506,13 +508,22 @@ function findMarketplacePlugin( state, pluginSlug, billingPeriod = '' ) {
+	// that represent a free plan as a non null cart item. This is why we check for both cases here. When we
+	// conduct a more thorough investigation and determine that PLAN_FREE is no longer, in fact, used to represent
+	// free plans in signup/onboarding, we can remove PLAN_FREE check.
+	return cartItems?.find( ( item ) => isPlan( item ) || item.product_slug === PLAN_FREE ) ?? null;
```
I understand you are creating the assertion this way to match the tests, but that doesn't feel right. It should be the other way. Right now, this doesn't make sense: "cart item exists if it's a plan or has a PLAN_FREE slug"? If I'm even reading this correctly 🤔

jeyip replied:
I completely understand the concern here -- I actually had a relatively long discussion with @jdc91  about this yesterday p1695244219714599-slack-CV2TX2PAN. Ultimately, although I'm somewhat confident we don't need the check and that I could simply adjust the tests, we both chatted about including this condition. Two things that convinced me were wanting to handle this part of the codebase with extra caution ( especially since it deals with checkout ) and the fact that, although free plans appear to be represented as a `null` value here, `PLAN_FREE` _is_ still used meaningfully outside of the context of `signup/onboarding`.

I'd much rather take my time in an isolated follow-up PR to verify that `item.product_slug === PLAN_FREE` isn't actually needed ( which I could set up immediately afterwards or in parallel if y'all agree with this )

ddc22 replied:
> I understand you are creating the assertion this way to match the tests

Actually this isn't the case, this line exactly describes the current state of the code base. We have two equally prevalent ways that define a "Free" Plan throughout the code. One is with `null or just the absence of a plan` and other is with a `plan_free ` product slug. Sorting this out will be its own investigative/refactoring task. I think you should explicitly mention this on top of that condition @jeyip.

jeyip replied:
> Actually this isn't the case, this line exactly describes the current state of the code base. We have two equally prevalent ways that define a "Free" Plan throughout the code. One is with null or just the absence of a plan and other is with a plan_free  product slug.

To add more nuance here, yes free plans can be represented as a product with a slug. For the scope of this specific portion of the codebase in `signup/plans`, and more specifically, the `Plans.onSelectPlan` method, it appears as if the free plan is only actually ever represented as a `null` value ( but I'm not 100% certain here ). This is why I think the test is outdated https://github.com/Automattic/wp-calypso/pull/81340/commits/e8c67368b5dc744f08bc73a688437055b9e1d362#diff-ba08cc37a1add3267ef49892c19f2073bba57ba096402e613fc4aa2a93aa32e6L52 

I'm happy to add a more detailed note about the current state of the codebase though 👍 

I haven't verified this myself yet, but I believe this to be important because if, theoretically, a non null free plan cart item were provided, I think the onboarding experience would probably behave unexpectedly and redirect the user to a checkout screen ( in spite of the fact that they're dealing with a free plan ). Either way, this specific condition that I've added maintains parity with what already exists in production, and I agree with Dehan that this should be a separate initiative.

ddc22 replied:
@jeyip Maybe you could actually write a small P2 about this complexity so that we have this documented on mgs. I think you could just copy paste the above 🙂

jeyip replied:
> Maybe you could actually write a small P2 about this complexity so that we have this documented on mgs. I think you could just copy paste the above 🙂

Ah this is a great idea 🙏  I'll draft something now

_edit:_ p4TIVU-aLF-p2

jeyip committed:
Fix case for selectedStorageOptionForPlans type

jeyip committed:
Move add-on transaction filter into block scope

jeyip committed:
Disable billing transaction query in onboarding

jeyip committed:
Fix upgrade click storage add-on data retrieval

jeyip committed:
Add named tuple destructuring for clearer tests

jeyip committed:
Update inline comment about check for plan free slug

jeyip committed:
DRY getPlanCartItem

jeyip commented:
@chriskmnds + @southp I've managed to tidy up a lot of code that I've been meaning go back to. This PR should mostly be ready for review.

I'm still thinking about one more outstanding question -- how do we best represent free plans in checkout code 🤔 Currently, it's a `null` value, which might not be ideal. Perhaps it _should_ be an empty array now? It seems like that could simplify checkout logic.

I'm holding off for now because the current logic seems to work perfectly well for us, and I don't foresee it being a problem in the future, but it's an easy change if folks have strong opinions.

jeyip committed:
Handle cartItems in stepper flows

jeyip commented on client/landing/stepper/declarative-flow/internals/steps-repository/plans/plans-wrapper.tsx:
```diff
@@ -92,11 +93,11 @@ const PlansWrapper: React.FC< Props > = ( props ) => {
+		if ( planCartItem ) {
 			recordTracksEvent( 'calypso_signup_plan_select', {
-				product_slug: selectedPlan?.product_slug,
-				free_trial: selectedPlan?.free_trial,
```
I remove the `free_trial` property because, according to tracks registration, it's not a valid event prop for `calypso_signup_plan_select`. https://github.com/Automattic/tracks-events-registration/blob/master/events/calypso_signup_plan_select.event.json

southp replied:
Great finding! Thanks for digging deep into it. There is also a deprecation note from the backend, so I believe it's safe to remove: 31d84-pb.

southp reviewed (commented):
@jeyip There is a minor Typing error :) I'm still half-way through the code, but would like to remind you first in case you missed it.
<img width="1280" alt="image" src="https://github.com/Automattic/wp-calypso/assets/1842898/bbff6bc5-7445-40bc-881c-9481061d4040">

southp commented on client/signup/config/flows.js:
```diff
@@ -32,7 +32,7 @@ function getCheckoutUrl( dependencies, localeSlug, flowName ) {
 
 function dependenciesContainCartItem( dependencies ) {
-	return dependencies.cartItem || dependencies.domainItem;
+	return dependencies.cartItem || dependencies.domainItem || dependencies.cartItems;
```
I think it's a good idea to keep `cartItem` here per this PR, but let's add an inlined TODO note, so folks won't be confused before we get to clean it up :)

southp commented on client/signup/main.jsx:
```diff
@@ -95,7 +96,10 @@ import './style.scss';
-	return !! ( dependencies && ( dependencies.cartItem || dependencies.domainItem ) );
+	return !! (
+		dependencies &&
+		( dependencies.cartItem || dependencies.domainItem || dependencies.cartItems )
```
Same as above. Let's add a TODO note regarding deprecation of cartItem.

southp commented on client/signup/main.jsx:
```diff
@@ -481,11 +485,15 @@ class Signup extends Component {
 		);
 		const hasCartItems = dependenciesContainCartItem( dependencies );
 		const cartItem = get( dependencies, 'cartItem' );
```
Another deprecation note can go here :)

jeyip committed:
Fix stepper onSelectPlan type error

southp commented:
@jeyip I still have a few pending cases to test from my test plan: https://github.com/Automattic/wp-calypso/pull/81340#issuecomment-1728669254, but would like to raise that `/start/domain` is broken. I haven't had a chance to dig into it yet, but my intuition is that it might be just a dependency configuration change.

southp reviewed (changes requested):
Thanks for the amazing collaboration on polishing this PR, @chriskmnds @jdc91 @jeyip . The code now looks in a much better shape and is very close. The final bit we need to address is the regression of the `/start/domain` flow. Once that is fixed and the deprecation inlined notes are added, this will be ready to launch :)

chriskmnds commented on client/my-sites/plan-features-2023-grid/index.tsx:
```diff
@@ -100,6 +102,7 @@ export interface PlanFeatures2023GridProps {
 	planTypeSelectorProps: PlanTypeSelectorProps;
 	// temporary: callback ref to scroll Odie AI Assistant into view once "Compare plans" button is clicked
 	observableForOdieRef: ( observableElement: Element | null ) => void;
+	selectedStorageOptions?: SelectedStorageOptionForPlans;
```
`selectedStorageOptions` prop doesn't seem relevant here and should be removed.
The only relevant type for your case is `PlanFeatures2023GridType` below I think

(it's a little confusing, I know. parts that are being cleaned up in other work)

jeyip replied:
Thanks for pointing this out! I removed the unnecessary type here https://github.com/Automattic/wp-calypso/pull/81340/commits/d5a8c5cf243569be2faf685300163ad81c12fd84

chriskmnds commented on client/my-sites/plan-features-2023-grid/index.tsx:
```diff
@@ -111,6 +114,7 @@ interface PlanFeatures2023GridType extends PlanFeatures2023GridProps {
 	isPlanUpgradeCreditEligible: boolean;
 	// temporary: element ref to scroll comparison grid into view once "Compare plans" button is clicked
 	plansComparisonGridRef: ForwardedRef< HTMLDivElement >;
+	selectedStorageOptions?: SelectedStorageOptionForPlans;
```
so this is fine and the only place you need to define this

jeyip committed:
Fix domain flow redirect error

There was a cartItem dependency for the site-picker step that needed to be updated to cartItems. On closer inspection, however, it appears as if the site-picker step doesn't need the dependency at all. The logic within the site-picker component makes no indication of needing it, so I remove the cartItem dependency instead.

Additionally, the plans step in the domain flow was being skipped under certain circumstances, so we update the provided cartItem dependency to cartItems in those circumstances as well.

jeyip commented:
> The final bit we need to address is the regression of the /start/domain flow.

There was a cartItem dependency for the `site-picker` step in the `domain` flow that needed to be updated to cartItems. On closer inspection, however, it appears as if the `site-picker` step doesn't need the dependency at all. The logic within the site-picker component makes no indication of needing it, so I remove the cartItem dependency instead. Some smoke testing seems to confirm this.

Additionally, the `plans` step in the `domain` flow was being skipped under certain circumstances, so we update the provided cartItem dependency to cartItems in those circumstances as well.

jeyip committed:
Remove unnecessary selectedStorageOption type

jeyip committed:
Add inline TODO note about deprecation of cartItem

southp reviewed (approved):
Beautifully done. Thanks for the great effort you've put into the fix, @jeyip . It's impressive to see you go extra miles further tidying the code up while fixing it up.

southp commented:
Also, since @chriskmnds has two PRs based on this PR(https://github.com/Automattic/wp-calypso/pull/81960 and https://github.com/Automattic/wp-calypso/pull/82054), I'm deploying this shortly to move things forward :)
